### ISSUE TITLE
Watir Default timeout is not working for button() method

### ISSUE THREAD

**Issue by Rajagopalan-M** [watir/watir#648]
WATIR default timeout is not working when I invoke button() method, I have written the following code browser.button(:value,"I Agree").click, sometime this line is reached even before page loads properly, and it throws the error immediately without waiting for given amount of time(30 seconds)

**Comment by titusfortner** [comment]
@Rajagopalan-M Can you verify the version you are using?

**Comment by Rajagopalan-M** [comment]
@titusfortner It's WATIR 6.8.0

![capture](https://user-images.githubusercontent.com/31696622/30218962-31c34862-94d8-11e7-863e-25502fdeba13.PNG)

**Comment by jkotests** [comment]
This can also be seen in 6.8.3. It seems to be problem whenever the button does not exist. Example simple do:

~~~~~~~~ruby
browser.button(id: 'does_not_exist').click
~~~~~~~~

It does wait in some conditions - eg button is disabled. Taking @jrmartinez03 example from #636:

~~~~~~~~ruby
url = 'https://app.launchdarkly.com/login'
browser = Watir::Browser.start(url)
Watir.default_timeout = 3

browser.button(css: ".Button:enabled").click #does not wait, immediately raises UnknownObjectException
browser.button(css: ".Button").click #does wait, eventually timing out with ObjectDisabledException
~~~~~~~~

**Comment by jkotests** [comment]
In [#wait_for_enabled](https://github.com/watir/watir/blob/master/lib/watir/elements/element.rb#L537-L553), it seems strange that we only check if the element exists when not an Input, Button, Select or Option. 
~~~~~~~~ruby
case self
when Input, Button, Select, Option
  # noop
else
  wait_for_exists && return
~~~~~~~~

Shouldn't we always being wait for the element to exist before checking that it is enabled? So something like:

~~~~~~~~ruby
wait_for_exists
return unless [Input, Button, Select, Option].any? { |klass| self.kind_of?(klass) }
~~~~~~~~

**Comment by titusfortner** [comment]
The idea is to try to minimize the number of wire calls for each thing. If `wait_until(&:enabled?)` already has to wait until exists, then this code prevents it from happening twice when it doesn't need to. But if it only assert exists first, then that doesn't help. :)

In this case, `#wait_for_exists` should only add one extra wire call, so it should be fine to pull it out of the case statement. `Watir::Executor` in Watir 7 should allow all of this to be more granularly controlled and should tidy our code quite a bit.

**Comment by jkotests** [comment]
What is the reason that `#enabled?` and `#disabled?` just `assert_exists` rather than `wait_for_exists` like other element methods?

In other words, why do we do:

~~~~~~~~ruby
def enabled?
  element_call(:assert_exists) { @element.enabled? }
end
~~~~~~~~

Rather than:

~~~~~~~~ruby
def enabled?
  element_call { @element.enabled? }
end
~~~~~~~~

**Comment by titusfortner** [comment]
When I first implemented `#element_call` there was an issue with how nesting calls would reset the timer and not time out. So `#visible?` and `#enabled?` needed special treatment. Maybe I just didn't change `#enabled?` when I could have after I fixed that. Do you have time to PR the change and see if it works?

**Comment by jkotests** [comment]
Both of the suggested solutions addressed the reported bug without breaking any of the existing tests. However, the change also slowed down the specs quite a bit:
* Without any changes: 8.8 minutes
* Switching `#visible?` and `#enabled?` to use `wait_for_exists`: 11.2 minutes
* Switching `wait_for_enabled` to always `wait_for_exists`: 10.2 minutes

That seems like too much of a performance hit.

**Comment by titusfortner** [comment]
Oh weird, it shouldn't be that much...
Firefox or Chrome?

**Comment by jkotests** [comment]
I was using Chrome 60.0.3112.113 on Windows 7 32-bit.

**Comment by titusfortner** [comment]
I was hoping it was Firefox because I've been seeing weird timing issues on that.
Any chance you can get me `Watir.logger.level = :debug` and `Selenium::WebDriver.logger.level = :info` for a spec or two that is related? I can try to look at it over the weekend.

**Comment by jkotests** [comment]
Maybe it's just my machine. I tried a third run without any of my changes... it took 20 minutes! I'll create a pull request for the change and we can see if Travis CI finds it any slower. If it is slower there too, I'll see if I can generate a useful log.

**Comment by titusfortner** [comment]
Ok, this is fixed in 6.8.4. Thanks @jkotests for your help figuring this out.

**Comment by Rajagopalan-M** [comment]
@titusfortner Thank you It works now
